File: examples/js/pvp-combat.ts
/**
 * Example: PvP challenge and turn-based combat (module: pvp_combat).
 *
 *   NEBULA_CONTRACT_ID=C... npx tsx pvp-combat.ts
 *
 * Two local accounts challenge each other and alternate moves until the
 * combat ends. Both must sign their own moves.
 */
import { Keypair } from "@stellar/stellar-sdk";
import { arg, config, fundWithFriendbot, invoke, loadPlayer, requireContractId, view } from "./lib/contract";

interface CombatState {
  combat_id: bigint;
  player1: string;
  player2: string;
  player1_hp: number;
  player2_hp: number;
  player1_energy: number;
  player2_energy: number;
  turn: string;
  status: string;
  winner?: string;
}

async function main() {
  requireContractId(config.contractId, "NEBULA_CONTRACT_ID");
  const orion = await loadPlayer();
  const vega = Keypair.random();
  await fundWithFriendbot(vega.publicKey());
  const players: Record<string, Keypair> = {
    [orion.publicKey()]: orion,
    [vega.publicKey()]: vega,
  };

  // Step 1: challenge (stake 0 = friendly match).
  const challengeId = await invoke<bigint>(orion, "create_challenge", [
    arg.address(orion.publicKey()),
    arg.address(vega.publicKey()),
    arg.i128(0),
  ]);
  console.log(`Challenge #${challengeId} sent`);

  // Step 2: the opponent accepts, which starts the combat.
  const combatId = await invoke<bigint>(vega, "accept_challenge", [
    arg.address(vega.publicKey()),
    arg.u64(challengeId),
  ]);
  console.log(`Combat #${combatId} started`);

  // Step 3: alternate moves. Read state (free), then the player on turn signs.
  for (let round = 1; round <= 40; round++) {
    const s = await view<CombatState>(orion, "get_combat", [arg.u64(combatId)]);
    if (s.status !== "active") break;

    const actor = players[s.turn];
    const energy = s.turn === s.player1 ? s.player1_energy : s.player2_energy;
    // Simple strategy: attack when there's enough energy, otherwise defend.
    const move = energy >= 15 ? "attack" : "defend";
    const power = move === "attack" ? 30 : 0;

    await invoke(actor, "execute_combat_move", [
      arg.address(actor.publicKey()),
      arg.u64(combatId),
      arg.symbol(move),
      arg.u32(power),
    ]);
    console.log(`  round ${round}: ${move}(${power})`);
  }

  // Step 4: results and ratings.
  const final = await view<CombatState>(orion, "get_combat", [arg.u64(combatId)]);
  console.log(`Status: ${final.status}, winner: ${final.winner ?? "none"}`);
  for (const kp of [orion, vega]) {
    const elo = await view<number>(orion, "get_elo_rating", [arg.address(kp.publicKey())]);
    console.log(`  ${kp.publicKey().slice(0, 6)}… ELO ${elo}`);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/pvp_combat.rs
//! # Example: PvP challenges, turn-based combat and ELO
//!
//! Module: `pvp_combat` via `NebulaNomadContract`
//!
//! Shows challenge -> accept -> alternate moves until a winner is found,
//! then reads the updated ELO ratings and combat history.
//!
//! Run:
//! ```text
//! cargo run --example pvp_combat
//! ```

use soroban_sdk::{symbol_short, testutils::Address as _, Address, Env};
use stellar_nebula_nomad::{NebulaNomadContract, NebulaNomadContractClient, INITIAL_ELO};

fn main() {
    let env = Env::default();
    env.mock_all_auths();
    let client = NebulaNomadContractClient::new(&env, &env.register(NebulaNomadContract, ()));

    let admin = Address::generate(&env);
    let orion = Address::generate(&env);
    let vega = Address::generate(&env);
    client.set_pvp_admin(&admin);

    println!("Starting ELO: {INITIAL_ELO}");

    // ── Step 1: Challenge ────────────────────────────────────────────────
    // Stakes are optional (0 = friendly match).
    let challenge_id = client.create_challenge(&orion, &vega, &0i128);
    println!("Challenge #{challenge_id}: Orion -> Vega");

    // ── Step 2: Accept (starts the combat) ───────────────────────────────
    let combat_id = client.accept_challenge(&vega, &challenge_id);
    let combat = client.get_combat(&combat_id);
    println!(
        "Combat #{combat_id} started: HP {} vs {}, first turn: {}",
        combat.player1_hp,
        combat.player2_hp,
        if combat.turn == orion { "Orion" } else { "Vega" }
    );

    // ── Step 3: Take turns ───────────────────────────────────────────────
    // Moves: "attack" deals `power` damage and costs power/2 energy.
    // "defend" restores energy. Only the player whose turn it is may move.
    for round in 1..=40 {
        let state = client.get_combat(&combat_id);
        if state.status != symbol_short!("active") {
            break;
        }
        let actor = state.turn.clone();
        let energy = if actor == state.player1 { state.player1_energy } else { state.player2_energy };

        // Simple strategy: attack when there's enough energy, else defend.
        let (mv, power) = if energy >= 15 {
            (symbol_short!("attack"), 30u32)
        } else {
            (symbol_short!("defend"), 0u32)
        };

        match client.try_execute_combat_move(&actor, &combat_id, &mv, &power) {
            Ok(Ok(())) => {
                let s = client.get_combat(&combat_id);
                println!(
                    "  round {round:>2}: {:<5} {:?}({power}) -> HP {} / {}",
                    if actor == orion { "Orion" } else { "Vega" },
                    mv,
                    s.player1_hp,
                    s.player2_hp
                );
            }
            Err(Ok(e)) => {
                println!("  round {round:>2}: move rejected: {e:?}");
                break;
            }
            other => {
                println!("  round {round:>2}: unexpected {other:?}");
                break;
            }
        }
    }

    // ── Step 4: Results ──────────────────────────────────────────────────
    let final_state = client.get_combat(&combat_id);
    match &final_state.winner {
        Some(w) if *w == orion => println!("Winner: Orion"),
        Some(_) => println!("Winner: Vega"),
        None => println!("No winner yet (status {:?})", final_state.status),
    }

    for (name, who) in [("Orion", &orion), ("Vega", &vega)] {
        let stats = client.get_combat_stats(who);
        println!(
            "{name:<5} ELO={} W/L/D={}/{}/{}",
            client.get_elo_rating(who),
            stats.wins,
            stats.losses,
            stats.draws
        );
    }

    let history = client.get_player_combat_history(&orion, &10u32);
    println!("Orion has {} recorded combat(s)", history.len());
}
